### ISSUE TITLE
Improve Mercurius test stability

### ISSUE THREAD

**Issue by jonnydgreen** [mercurius-js/mercurius#476]
As suggested in this conversation https://github.com/mercurius-js/mercurius/pull/470#issuecomment-828245464, I'm opening this issue to discuss improvements to the Mercurius test stability before refactoring them to be more stable.

## Observed issues

| Test                                                                                                                                                                                                                  | Environment(s)                 |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------- |
| [`gateway/pollingInterval.js: Polling schemas (subscriptions should be handled)`](#polling-schemas-subscriptions-should-be-handled)                                                                                   | `[windows-latest]-[14.x.x]` |
| [`gateway/remote-services.js: All tests`](#remote-services.js)                                                                                                                                                        | All tests                   |
| [`gateway/pollingInterval.js: Polling schemas (should properly regenerate the schema when a downstream service restarts)`](#polling-schemas-should-properly-regenerate-the-schema-when-a-downstream-service-restarts) | All tests                   |

### Polling schemas (subscriptions should be handled)

This is easily seen in the CI if you rewrite the tests to repeat until failure (I have tested this on a test branch on my fork). From my initial experiments, this test fails ~33% of the time. I am yet to see this on a local machine (MacOS), however I am loading mercurius onto a Windows PC asap to replicate the exact environment as closely as possible.

The point of failure is [this line here](https://github.com/mercurius-js/mercurius/blob/master/test/gateway/pollingInterval.js#L1179). None of the assertions after this line are reached, and from the logs, the test is exited prematurely (causing a test plan error). I need to dig deeper to find the exact cause, but my initial thoughts are that since it is intermittent, it is some race condition.

### remote-services.js

The tests in this file take about ~8s to run which suggests that everything is not being cleaned up. From an initial look at the code, I think it might be to do with undici connections not being cleaned up when `buildFederation` fails. If this is correct, this is a fairly simple fix, but tricky to test!

### Polling schemas (should properly regenerate the schema when a downstream service restarts)

This test takes ~4s to run which again suggests that everything is not being cleaned up properly. Similarly to the `remote-services.js` being slow, this looks like it may be to do with the undici clients not being closed in the gateway when a downstream service is closed, however, I need to dig deeper to confirm the exact cause!

@mcollina @PabloSzx @PacoDu feel free to add any other issues you have seen in the Mercurius tests to this issue so we can fix everything at once in a related PR! I can open a draft PR asap once I've determined solutions for some of the above :)

**Comment by mcollina** [comment]
Go for it!!

**Comment by jonnydgreen** [comment]
Quick update! I've tracked down the issue in `Polling schemas (subscriptions should be handled)` - this was due to the `SubscriptionClient` not being ready when a subscription is created. The reason this was happening only occasionally was due to the fact that the `connection_init` and `start` were sent at the same time. I've got a potential fix which is looking promising from my testing. I'll submit a PR for these ASAP including fixes for making tracking down problems like easier in the future! :)
